cabal-install/Distribution/Client/ProjectOrchestration.hs

```diff
@@ -466,16 +472,135 @@ printPlan verbosity
     showMonitorChangedReason  MonitorCorruptCache = "cannot read state cache"
 
 
-reportBuildFailures :: BuildResults -> IO ()
-reportBuildFailures plan =
+reportBuildFailures :: ElaboratedInstallPlan -> BuildResults -> IO ()
+reportBuildFailures plan buildResults
+  | null failures
+  = return ()
 
-  case [ (pkgid, reason)
-       | (pkgid, Left reason) <- Map.toList plan ] of
-    []      -> return ()
-    _failed -> exitFailure
-    --TODO: [required eventually] see the old printBuildFailures for an example
-    -- of the kind of things we could report, but we want to handle the special
-    -- case of the current package better, since if you do "cabal build" then
-    -- you don't need a lot of context to explain where the ghc error message
-    -- comes from, and indeed extra noise would just be annoying.
+  | isSimpleCase
+  = exitFailure
+
+  | otherwise
+  = case failuresPrimary of
+     [(pkg, reason)] -> die $ renderFailure pkg reason
+     multiple        -> die $ "multiple failures:\n"
+                           ++ unlines
+                                [ renderFailure pkg reason
+                                | (pkg, reason) <- multiple ]
+  where
+    failures =  [ (pkgid, reason)
+                | (pkgid, Left reason) <- Map.toList buildResults ]
+
+    failuresPrimary =
+      [ (pkg, reason)
+      | (pkgid, reason) <- failures
+      , case reason of
+          DependentFailed {} -> False
+          _                  -> True
+      , InstallPlan.Configured pkg <-
+           maybeToList (InstallPlan.lookup plan pkgid)
+      ]
+
+    -- Special case: we don't want to report anything complicated in the case
+    -- of just doing build on the current package, since it's clear from
+    -- context which package failed.
+    --
+    -- We generalise this rule as follows:
+    --  - if only one failure occurs, and it is in a single root package (ie a
+    --    package with nothing else depending on it)
+    --  - and that failure is of a kind that always reports enough detail
+    --    itself (e.g. ghc reporting errors on stdout)
+    --  - then we do not report additional error detail or context.
+    --
+    isSimpleCase
+      | [(pkgid, reason)] <- failures
+      , [pkg]             <- rootpkgs
+      , installedUnitId pkg == pkgid
+      , isFailureSelfExplanatory reason
+      = True
+      | otherwise
+      = False
+
+    isFailureSelfExplanatory (BuildFailed e)
+      | Just (ExitFailure _) <- fromException e = True
+
+    isFailureSelfExplanatory (ConfigureFailed e)
+      | Just (ExitFailure _) <- fromException e = True
+
+    isFailureSelfExplanatory _                  = False
+
+    rootpkgs =
+      [ pkg
+      | InstallPlan.Configured pkg <- InstallPlan.toList plan
+      , hasNoDependents pkg ]
```

Comment:
Note that root is a slight misnomer: we think of user explicitly specified build targets as roots, but if a user asks for Cabal and cabal-install, Cabal would not be considered a root.